src/mds/SessionMap.h

```diff
@@ -215,20 +216,54 @@ class Session : public RefCountedObject {
     ceph_assert(!info.prealloc_inos.empty());
 
     if (ino) {
-      if (info.prealloc_inos.contains(ino))
+      if (info.prealloc_inos.contains(ino)) {
 	info.prealloc_inos.erase(ino);
-      else
+	if (delegated_inos.contains(ino))
+	  delegated_inos.erase(ino);
+      } else {
 	ino = 0;
+      }
     }
     if (!ino) {
-      ino = info.prealloc_inos.range_start();
-      info.prealloc_inos.erase(ino);
+      /* Grab first prealloc_ino that isn't delegated */
+      for (interval_set<inodeno_t>::const_iterator p = info.prealloc_inos.begin(); p != info.prealloc_inos.end(); ++p) {
+	for (auto i = (*p).first ; i < (*p).first+(*p).second ; i += 1) {
+	  if (!delegated_inos.contains(i)) {
+	    ino = i;
+	    info.prealloc_inos.erase(ino);
+	    break;
+	  }
+	}
+	if (ino)
+	  break;
+      }
     }
     info.used_inos.insert(ino, 1);
     return ino;
   }
+  void delegate_inos(int want) {
+    want = std::min((int)info.prealloc_inos.size()/2, want - (int)delegated_inos.size());
+    if (want <= 0)
+      return;
+
+    for (interval_set<inodeno_t>::const_iterator p = info.prealloc_inos.begin(); p != info.prealloc_inos.end(); ++p) {
+	for (auto i = (*p).first ; i < (*p).first+(*p).second ; i += 1) {
+	  if (!delegated_inos.contains(i)) {
```

Comment:
same here

Reply:
maybe we can take  delegate_inos in reverse order, from prealloc_inos' end to start. this can make take_ino() more efficient

Reply:
Yeah, that may be possible. OTOH, if we use the ranges that the contains() helper returns, then iterating over them will already be pretty efficient. I'll consider it though.